### PR TITLE
New blockio system

### PR DESCRIPTION
**Pull request by jmaggio14**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -7,36 +7,169 @@ os:
...
-  # Building documentation
-  - sphinx-apidoc -o docs/source imagepypelines
-  - cd docs && make html
-  - cd ..
-  - cp CNAME docs/build/html/CNAME
+  # running doctest on website examples & code snippets
+  - py.test --cov=./imagepypelines --cov-append --doctest-glob="*.rst" --ignore=setup.py --ignore=tests/ --ignore=imagepypelies_gui/
+
+after_script:
+  - kill -9 $PREVENT_TRAVIS_TIMEOUT_PID
```
Comment (JoeBartelmo) [inline .travis.yml]: disgusting

File: .travis.yml [hunk 2]
```diff
@@ -7,36 +7,169 @@ os:
...
+  # print out opencv build script for full traceability
+  # - cat ~/opencv_build.txt
+  # - ls -R /home/travis/virtualenv/
+  # - ls -R /opt/python/"$TRAVIS_PYTHON_VERSION".7/include
 
 after_success:
+  # upload coverage report
   - codecov
+  # Building website
+  - cd docs && make html && cd ..
```
Comment (JoeBartelmo) [inline .travis.yml]: recommend pushd and popd over  cd and cd ..

File: docs/prevent_travis_timeout.sh [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# this file prints something to standard out every 5 minutes to prevent travis
+# from automatically ending the build. It can be stopped by killing it's PID
+while [ 1 ]; do
+    sleep 5m
+    echo "(IP TRAVIS) this message is printed to prevent the travis build from auto timing out"
```
Comment (JoeBartelmo) [inline docs/prevent_travis_timeout.sh]: this is hilarious

Comment (JoeBartelmo) [reply docs/prevent_travis_timeout.sh]: I miss your jank

File: docs/source/_static/js/copybutton.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
+$(document).ready(function() {
+    /* Add a [>>>] button on the top-right corner of code samples to hide
+     * the >>> and ... prompts and the output and thus make the code
+     * copyable. */
+    var div = $('.highlight-python .highlight,' +
```
Comment (JoeBartelmo) [inline docs/source/_static/js/copybutton.js]: try to use `let` over `var` in js, `const` if you're able to. var allows bad scope access that can mess with things elsewhere, doesn't matter too much here, but a good practice.

File: imagepypelines/builtin_blocks/FTP.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+# @github: https://github.com/jmaggio14/imagepypelines
+#
+# Copyright (c) 2018-2019 Jeff Maggio, Nathan Dileas, Ryan Hartzell
+#
+import ftplib
+import io
+from .. import SimpleBlock
+
+
+class FTP(SimpleBlock):
```
Comment (JoeBartelmo) [inline imagepypelines/builtin_blocks/FTP.py]: I'm assuming your block here is a file upload? I'd recommend adding a block for a REST call instead (or in addition). Plenty of frameworks will allow uploads VIA rest and can then process directly in their application.

File: imagepypelines/core/io/output.py [hunk 6]
```diff
@@ -97,3 +101,55 @@ def make_numbered_prefix(file_number,number_digits=5):
...
+    file_path, ext = os.path.splitext(fname)
+    if output_dir is None:
+        out_name = file_path + '.' + format
+    else:
+        basename = os.path.basename(file_path)
+        out_name = os.path.join(output_dir, basename + '.' + format)
+
+
+    if format not in IMAGE_EXTENSIONS:
+        raise TypeError("format must be one of {}".format(IMAGE_EXTENSIONS))
```
Comment (JoeBartelmo) [inline imagepypelines/core/io/output.py]: check should happen after line 126.

File: docs/source/_static/js/copybutton.js (hunk 4 above)
Comment (jmaggio14) [reply docs/source/_static/js/copybutton.js]: I didn't know those keywords even existed in js, thank you!

**Pull request merged**